### DISCUSSION TITLE
Does Grails need an IDE Plugin to reach its full potential?

### DISCUSSION THREAD

**Discussion by gsartori** [apache/grails-core#13536] [Polls]
The era of Pivotal STS with Grails support and JetBrains IntelliJ IDEA Grails plugin is over. They are investing respectively in Spring and Kotlin. This is not because their software is better than Groovy & Grails, it's just a business decision.

I think we should understand wether Grails can achieve its goals by itself on an IDE supporting just the Groovy language or if it needs its onwn Grails Plugin to reach its full potential. In this case, should we have a Grails plugin for the most used IDEs? 

We may also have Grails features included in the Groovy Language Plugins. There are already Groovy plugins available for the most used IDEs, it may be easier to build on them instead of starting from scratch. We don't need thousands of features, we just need the IDE to recognise the dynamic Grails features (params, render, injected traits, GroovyORM dynamic finders, etc).

We need to understand what Grails needs and what effort we can put on this so we can take a consious decision.

**Comment by Taack** [comment]
We heavily rely on intellij because of its groovy support. It would be cool if domain class could be supported by intellij oss version, and keep the ultimate version for GSP.

We replaced GSP by DSL. See https://taack.org .

**Reply by jamesfredley** [reply]
My thoughts on IDE Plugin support for Grails:  https://github.com/orgs/grails/discussions/13546#discussioncomment-10088884

**Comment by matrei** [comment]
> The era of Pivotal STS with Grails support and JetBrains IntelliJ IDEA Grails plugin is over.

It this a fact?
I'm using the Grails plugin in IntelliJ.

**Reply by gsartori** [reply]
They haven’t publicly declared it, so I guess it’s not a fact. But it’s a way of looking at things that may help us going forward. My opinion is that we need to invest in open source tools to make Grails “great again” (just to cite some random guy from the US, lol) and the IntelliJ IDEA Grails plugin is not.

**Comment by snimavat** [comment]
I dont think I need an IDE with support specifically for grails.
I use intellij community, import grails as a gradle project, Groovy support in intellij is all i need.
We dont use GSP, dynamic finders etc (99% code is compile static). 

I or any one else in our team dont miss "Grails IDE"


IDE support is not a blocker, I would rather prefer the dev time is spent some where else.

**Reply by gsartori** [reply]
My experience with junior developers is that they get confused if the IDE gets confused about code that is actually correct.

Also my opinion is that we should give support for features that are “core” to the framework. This happens with any major framework out there. In our case code recognition/completion for dynamic calls is the most impacting one regarding productivity and ergonomics. 

A Grails plugin can also help communicating the framework giving the idea that it is something widely used or adopted (or at least it gives the impression it is a “product”). It can be a marketing tool for the grails community. 

In my experience Senior developers are less impacted by the lack of a Grails plugin but they would be happy to have one as well.

**Comment by kico-itexto** [comment]
Plugins IDE are a MUST.

Grails was initially projected to be used on a simple code editor and a CLI, not depending on any IDE. And it works fine this way: but I really miss things like a basic code complete feature.

And as a Grails consultant I saw many times people NOT choosing Groovy and Grails because of the lack of IDE plugins. Just to remember, the era where Grails grew most was when we had GGTS maintened by Pivotal and the support that Nebeans offered for the project. And people was happy with it.

**Reply by gsartori** [reply]
I feel the same, my view on this is that adoption of Grails lowered also because people thought the framework was not supported/adopted anymore by big players, not because of the lack of features or the technology itself (even though I realise the Grails 3 developers experience was better than the current one mainly because of app reloading speed in my case).

When C level or managers take decisions on what technology their company will adopt they judge with different metrics respect to developers.

For instance I would prefer Micronaut from a development point of view but I ended up suggesting Spring to my company for service implementation since it is most widely adopted at the moment and the company I am working for is not big enough to be an early adopter.

**Reply by kico-itexto** [reply]
Yeah @gsartori , good point: there's always the FUD factor.

I see Grails as a perfect fit for many customers, but when I mention it the resistance is HUGE in part because we don't have a good support on IDEs and the strategy of dissemination of the framework is very timid. We used to have events about Groovy and Grails: we don't see many anymore. The same can be said about blog posts, videos and so on.

On Brazil I was basically the only one speaking about the framework: wrote a book about it (Falando de Grails) and also maintened until few months a go a large forum about it (Grails Brasil), which I had to terminate because basically no one else was publishing there.

We can see that there's development on the platform (we, deveopers which use Grails for a long time), but newbies have the perception that this is now a ghost town, which is a shame.
